### PR TITLE
feat: Indexed (and concurrent) Merkle tree for nullifiers

### PR DESCRIPTION
**Pull request by vadorovsky**

### CODE REVIEW

File: merkle-tree/concurrent/src/concurrent_merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            //
+            // * Valitating a proof.
+            // * Performing procedures which usually are done by `replace_leaf`
+            //   algorithm.
+            //
+            // Here, we just call compute the new root and call `replace_leaf`
+            // directly.
+            leaf = H::hashv(&[&leaf, &H::zero_bytes()[1]])?;
+            let proof = self.rightmost_proof.clone();
+            self.replace_leaf_inner(H::zero_bytes()[0], leaf, 0, 0, &proof)?;
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/concurrent_merkle_tree.rs:192]: we could also just start with index 1 instead and leave index zero the zero values

File: merkle-tree/concurrent/src/merkle_path.rs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#[repr(C)]
+pub struct MerklePath<const MAX_DEPTH: usize> {
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/merkle_path.rs:2]: MAX_DEPTH is equivalent with height merkle tree or MAX_HEIGHT?

File: merkle-tree/concurrent/src/concurrent_merkle_tree.rs (hunk 1 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/src/concurrent_merkle_tree.rs:198]: I'm not sure I'm following your idea.

`H::zero_bytes()[0]` is passed here as `old_leaf` parameter - this way we make it explicit that we replace the first leaf, being zero at that time, with a new one. Feels intuitive (at least for me :D) and I wouldn't like to obfuscate it.

But I also feel like I'm misunderstanding you. Which parameter would you like to replace in this call?

File: merkle-tree/concurrent/src/merkle_path.rs (hunk 2 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/src/merkle_path.rs]: The whole `MerklePath` stuff is not used anymore and I'm using `MAX_HEIGHT` consistently now. Thanks for pointing out!

File: merkle-tree/concurrent/src/concurrent_merkle_tree.rs (hunk 1 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/src/concurrent_merkle_tree.rs:198]: Ah, OK, I probably get it now. You mean if we could just leave the 1st leaf/subtree (index zero) always as zero byte? That would be inconsistent with spl-concurrent-merkle-tree, so we couldn't go forward with your idea of testing both implementations.

(But maybe I'm still misunderstanding you :P )

Comment (ananas-block) [reply merkle-tree/concurrent/src/concurrent_merkle_tree.rs:198]: I see makes sense.

**Pull request closed without merging**